src/coins.h

```diff
@@ -103,8 +106,31 @@ class Coin
  */
 struct CCoinsCacheEntry
 {
+private:
+    /**
+     * These are used to create a doubly linked list of flagged entries.
+     * They are set in AddFlags and unset in ClearFlags.
+     * A flagged entry is any entry that is either DIRTY, FRESH, or both.
+     *
+     * DIRTY entries are tracked so that only modified entries can be passed to
+     * the parent cache to do a batch write. This is a performance optimization
+     * compared to giving all entries in the cache to the parent and having the
+     * parent scan for only modified entries.
+     *
+     * FRESH entries are tracked because a FRESH-but-not-DIRTY entry can only be
+     * a spent coin that exists in the parent cache. When performing a batch
+     * write but not clearing the cache (CCoinsViewCache::Sync), we must also
+     * erase all spent entries. Since only DIRTY or FRESH entries can be spent,
+     * we can track them in the linked list to erase them. This is a performance
+     * optimization compared to scanning all cache entries to find any that are
+     * spent-but-not-DIRTY.
+     */
+    CoinsCachePair* m_prev{nullptr};
```

Comment:
I don't fully understand how we need `m_prev` here, but have you investigated whether it's possible to get rid of it and rely on `m_next` only?
Sometimes we still have a reference to both, but I assume that wasn't always the case, right?

Reply:
With `m_next` we have constant time insertion at the head, which would be fine if we only added flagged entries and cleared them as we traverse them.

However, we also need to delete entries in constant time if they are both `FRESH` and `DIRTY` and get spent. For this we need a doubly linked list.

Reply:
What prohibits us from accessing the previous ones in constant time during iteration (I haven't had the energy to delve into it properly yet)?

Reply:
> What prohibits us from accessing the previous ones in constant time during iteration

Nothing during iteration. But we need to delete entries randomly when a FRESH coin is spent, which is not during any iteration.

Reply:
This kept me up at night, we should be able to get rid of `m_prev` by doing insertions and deletions slightly differently:
* we'd have an immutable `SENTINEL` value always serving as the last element (instesd of `nullptr`);
* insertion would always be in front of the previous element, which would serve as the new head (iterating from which will always get us to the `SENTINEL`);
* deletion would copy the content (`node`, `flag` and `next`) of the next node instead (unless it's a `SENTINEL`), and dispose that instead.

Something like this:
```patch
diff --git a/src/coins.h b/src/coins.h
--- a/src/coins.h	(revision c36363f6b24c7ab2afe198d9855f507ddf096e1f)
+++ b/src/coins.h	(date 1719755615706)
@@ -123,11 +123,11 @@
      * parent cache, the FRESH-but-not-DIRTY coin will be tracked by the linked
      * list and deleted when Sync or Flush is called on the CCoinsViewCache.
      */
-    CoinsCachePair* m_prev{nullptr};
     CoinsCachePair* m_next{nullptr};
     uint8_t m_flags{0};
 
 public:
+    static CoinsCachePair SENTINEL;
     Coin coin; // The actual cached data.
 
     enum Flags {
@@ -154,22 +154,19 @@
     //! Adding a flag also requires a self reference to the pair that contains
     //! this entry in the CCoinsCache map and a reference to the head of the
     //! flagged pair linked list.
-    inline void AddFlags(uint8_t flags, CoinsCachePair& self, CoinsCachePair& head)
+    inline void AddFlags(uint8_t flags, CoinsCachePair& next)
     {
         if (!m_flags && flags) {
-            m_prev = &head;
-            m_next = head.second.m_next;
-            if (m_next) m_next->second.m_prev = &self;
-            head.second.m_next = &self;
+            m_next = &next;
         }
         m_flags |= flags;
     }
     inline void ClearFlags()
     {
-        if (m_flags) {
-            if (m_next) m_next->second.m_prev = m_prev;
-            m_prev->second.m_next = m_next;
-            m_prev = m_next = nullptr;
+        if (m_flags && m_next != &SENTINEL) {
+            coin = std::move(m_next->second.coin);
+            m_flags = m_next->second.m_flags;
+            m_next = m_next->second.m_next;
         }
         m_flags = 0;
     }
```